### ISSUE TITLE
- 'OrderedDict' object has no attribute 'read'

### ISSUE THREAD

**Issue by mooperd** [saltstack/salt#21935]
Could anyone tell me what the problem is here? This feels like a bug.

Thanks,

Andrew

```
[root@ip-10-16-1-46 pillar]# salt 'test-freeipa-client-2' pillar.items
test-freeipa-client-2:
    ----------
    _errors:
        - Rendering Primary Top file failed, render error:
        - 'OrderedDict' object has no attribute 'read'
[root@ip-10-16-1-46 pillar]# cat top.sls 
base:
#  'test-freeipa-*':
#    - freeipa
  'sensu-*':
    - sensu
  '*':
    - ip
```

salt --versions

```
           Salt: 2015.2.0
         Python: 2.7.5 (default, Jun 17 2014, 18:11:42)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.4.4
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.4
           Mako: Not Installed
```

/srv/pillar/top.sls

```
base:
#  'test-freeipa-*':
#    - freeipa
  'sensu-*':
    - sensu
  '*':
    - ip
```

salt-master --log-level=debug

```
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: ip-10-16-1-46.eu-west-1.compute.internal
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Including configuration from '/etc/salt/master.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/master.d/reactor.conf
[DEBUG   ] Installed boto version: 2.25.0
[DEBUG   ] LazyLoaded local_cache.clean_old_jobs
[DEBUG   ] Updating fileserver cache
[DEBUG   ] This salt-master instance has accepted 10 minion keys.
[INFO    ] Clear payload received with command publish
[DEBUG   ] Sending event - data = {'_stamp': '2015-03-24T14:04:03.947898', 'minions': ['test-freeipa-client-2']}
[DEBUG   ] Sending event - data = {'tgt_type': 'glob', 'jid': '20150324140403947564', 'tgt': 'test-freeipa-client-2', '_stamp': '2015-03-24T14:04:03.948152', 'user': 'sudo_centos', 'arg': [], 'fun': 'pillar.items', 'minions': ['test-freeipa-client-2']}
[INFO    ] User sudo_centos Published command pillar.items with jid 20150324140403947564
[DEBUG   ] Published command details {'tgt_type': 'glob', 'jid': '20150324140403947564', 'tgt': 'test-freeipa-client-2', 'ret': '', 'user': 'sudo_centos', 'arg': [], 'fun': 'pillar.items'}
[DEBUG   ] Gathering reactors for tag 20150324140403947564
[DEBUG   ] Gathering reactors for tag salt/job/20150324140403947564/new
[INFO    ] Clear payload received with command _auth
[INFO    ] Authentication request from test-freeipa-client-2
[INFO    ] Authentication accepted from test-freeipa-client-2
[DEBUG   ] Sending event - data = {'id': 'test-freeipa-client-2', '_stamp': '2015-03-24T14:04:04.027593', 'result': True, 'pub': '-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAvhx0xfXSVrOBVNhD2ItH\nql2JOm1LiRpSeNnfUa14Gf7V1f32eXExzy92tuohARToQal6Q1a23m4z2M7te57v\nBvIG5DXZ9nooOiN+CnlH7RC/8/VnzIMBuZfZmk5/UtnDv6slzSX5LY+Z2h4IVm9W\nEjytafjdzhMYA3/1v9fo9iPVbpuj+zDQWnNdvQGfxzUewAhCTGQ6wphLKYOWkN/g\nwXx123ZVQoMMXwR3QcmQRMkrb7Uu+wfqfcKhbcwhVOzGzASarLVczk8cBN8yJWBB\nyD/2Z/hMGxnDRuJmluX5DDpvktWasoP3gNNNhcTzGh1HkjuFVbRtwXu6Z14g33AK\nrkk8Cd/8s6wpaST/T0+2/hAl8rUu4y0fZtlWEYcai1E8xcIc17vWIXG7UWuawWC4\nz/Mp8ShQ03CkKlQqA5DeDvjmXc3KKfofdgCR9iC0LqfHusTsw2MC3fnFXogpVj3E\n/BClV+ZmpPxizPCRuGygROT6GPiRbj5XqEpY88z8XYXwrDRpTouubO4SzNxbpRxj\niZYKto+O5QKPN1N8vOv0TnafMrZBku9kvm/Px5TeQ+5sNhVsPdFcTVz8MLaxVeuD\npPrVx0UrhW7G1eSsLUpOC3ymMmZzcFt4GAG93NG2aiiduh6eWpbl3fMLT0mivRtq\nqx3aPgGn4PZSFqv30JNDZ0ECAwEAAQ==\n-----END PUBLIC KEY-----\n', 'act': 'accept'}
[DEBUG   ] Gathering reactors for tag salt/auth
[INFO    ] AES payload received with command _pillar
[DEBUG   ] Results of YAML rendering: 
OrderedDict([('base', OrderedDict([('sensu-*', ['sensu']), ('*', ['ip'])]))])
[CRITICAL] Pillar render error: Rendering Primary Top file failed, render error:
'OrderedDict' object has no attribute 'read'
[INFO    ] Clear payload received with command _auth
[INFO    ] Authentication request from test-freeipa-client-2
[INFO    ] Authentication accepted from test-freeipa-client-2
[DEBUG   ] Sending event - data = {'id': 'test-freeipa-client-2', '_stamp': '2015-03-24T14:04:04.213319', 'result': True, 'pub': '-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAvhx0xfXSVrOBVNhD2ItH\nql2JOm1LiRpSeNnfUa14Gf7V1f32eXExzy92tuohARToQal6Q1a23m4z2M7te57v\nBvIG5DXZ9nooOiN+CnlH7RC/8/VnzIMBuZfZmk5/UtnDv6slzSX5LY+Z2h4IVm9W\nEjytafjdzhMYA3/1v9fo9iPVbpuj+zDQWnNdvQGfxzUewAhCTGQ6wphLKYOWkN/g\nwXx123ZVQoMMXwR3QcmQRMkrb7Uu+wfqfcKhbcwhVOzGzASarLVczk8cBN8yJWBB\nyD/2Z/hMGxnDRuJmluX5DDpvktWasoP3gNNNhcTzGh1HkjuFVbRtwXu6Z14g33AK\nrkk8Cd/8s6wpaST/T0+2/hAl8rUu4y0fZtlWEYcai1E8xcIc17vWIXG7UWuawWC4\nz/Mp8ShQ03CkKlQqA5DeDvjmXc3KKfofdgCR9iC0LqfHusTsw2MC3fnFXogpVj3E\n/BClV+ZmpPxizPCRuGygROT6GPiRbj5XqEpY88z8XYXwrDRpTouubO4SzNxbpRxj\niZYKto+O5QKPN1N8vOv0TnafMrZBku9kvm/Px5TeQ+5sNhVsPdFcTVz8MLaxVeuD\npPrVx0UrhW7G1eSsLUpOC3ymMmZzcFt4GAG93NG2aiiduh6eWpbl3fMLT0mivRtq\nqx3aPgGn4PZSFqv30JNDZ0ECAwEAAQ==\n-----END PUBLIC KEY-----\n', 'act': 'accept'}
[DEBUG   ] Gathering reactors for tag salt/auth
[INFO    ] AES payload received with command _return
[INFO    ] Got return from test-freeipa-client-2 for job 20150324140403947564
[DEBUG   ] Sending event - data = {'fun_args': [], 'jid': '20150324140403947564', 'return': {'_errors': ["Rendering Primary Top file failed, render error:\n'OrderedDict' object has no attribute 'read'"]}, 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2015-03-24T14:04:04.278890', 'fun': 'pillar.items', 'id': 'test-freeipa-client-2'}
[DEBUG   ] Gathering reactors for tag salt/job/20150324140403947564/ret/test-freeipa-client-2
```

**Comment by basepi** [comment]
Hmm, that is quite strange. Could you try removing the commented out lines? You know, for science? I can't imagine they'd be effecting anything, but the rest of the topfile looks just fine...

Also, how did you install salt? Since 2015.2.0 isn't out yet, it should be reporting the SHA of the installed version, but for certain install types that can be lost.

**Comment by jfindlay** [comment]
@mooperd, I can't reproduce your error with the pillar top file you've provided.  I'm not sure what's going on here.

**Comment by mooperd** [comment]
I've been messing around with salt master config a bit to get the gpg parser working. This system is perhaps now a bit old and confused.  I'll redeploy everything and see how it works out.
Ta,
Andrew
On 24 Mar 2015 20:56, jfindlay <notifications@github.com> wrote:@mooperd, I can't reproduce your error with the pillar top file you've provided.  I'm not sure what's going on here.

**Comment by cachedout** [comment]
@mooperd Any further information on this issue? How did the redeploy go?

**Comment by mooperd** [comment]
I wasn't able to find time to do more with this issue. Sorry about that.

**Comment by justindesilets** [comment]
I can confirm this issue is related to the GPG renderer if it is in a bad state. I came across this post when researching the error called out in this issue: https://github.com/saltstack/salt/issues/38616 . I tried hacking my way around the gpg issue and then updated my salt master's  renderer setting to be: `renderer: gpg | yaml | jinja` after that I kept getting the above error. I then commented out the settings and instead tried using a `#!yaml|gpg` statement at the top of my pillar file. The one block of GPG text was not rendered. I am now researching why I am getting the `gpg: can't connect to the agent: IPC connect call failed
gpg: problem with the agent: No agent running
gpg: can't connect to the agent: IPC connect call failed
gpg: problem with the agent: No agent running
gpg: Key generation canceled.`
issue. This is on a cleanly built CentOS 7.3 VM where gnupg was installed by default. Version `gnupg2-2.0.22-4.el7.x86_64`. If I find a fix I will update again so it can be added into the documentation.
